[ShapeOpt] Adding bead optimization

dbaumgaertner opened the pull request:
This PR introduces:

- A new algorithm to perform bead optimization using an Augmented Lagrange approach. The algorithm allows specifying the bead height as well as the bead direction (extrusion on only one side, on both sides, in the local normal direction or in a specified direction) Additionally if specified, the algorithm considers a built-in penalty term which enforces a binary distribution in the control space. Damping and a direct constraint enforcement are included to allow a transition to non-design areas.

- A value logger that logs the results of a bead optimization in detail. 
- A constant filter function
- A unit test for bead optimization

Additionally, the following modifications were introduced:

- The structure of the algorithms slightly changed such that each algorithm introduces its own set of variables to (to avoid an overflow of nodal variables) 

- The output of all value loggers is cleaned

- The application of the variables TIME / STEP in the analyzer and the mesh solver is corrected. I.e., from the optimizer point of view, the latter variables are not modified anymore. Advancing in time is exclusively controlled by the individual optimization algorithms. In doing so, historical nodal optimization variables may be properly used within an algorithm, which was not the case before.

- The settings validation in the mesh solver was corrected such that mesh velocities are not computed anymore (creates problems due to the variable DELTA_TIME which is meaningless and not needed in optimization)

- The import of the model part is moved to the initialize function of each algorithm. By that, the entire workflow is defined and readable from the algorithm. This shall facilitate the process understanding.

dbaumgaertner committed:
introduced default gid config settings

this avoids unnecessary code duplication with test examples and hides mostly constant settings

dbaumgaertner committed:
Merge branch 'fix-gid-output-in-shape-opt' into feature-corrected-bead-optimization

dbaumgaertner committed:
introduced auxiliary nodal variables

dbaumgaertner committed:
introduced plate as test example

dbaumgaertner committed:
introduced draft for algorithm for bead optimization

dbaumgaertner committed:
introduced draft for value logging with new algorithm

dbaumgaertner committed:
Merge branch 'feature-generalize-vertex-morphing-mapping' into feature-corrected-bead-optimization

dbaumgaertner committed:
Merge branch 'feature-generalize-vertex-morphing-mapping' into feature-corrected-bead-optimization

dbaumgaertner committed:
clean optimization algorithm and introduced convergence checks

dbaumgaertner committed:
minor change in settings to include tolerance check

dbaumgaertner committed:
improved output with additional nodal variables

maybe deleted again later

dbaumgaertner committed:
improved output of all value logger

dbaumgaertner committed:
first working and cleaned implementation

dbaumgaertner committed:
Merge branch 'master' into feature-corrected-bead-optimization

dbaumgaertner committed:
cleaned algorithms and their variable handling

with this commit each algorithm individually adds variables as needed

dbaumgaertner committed:
included some options

dbaumgaertner committed:
added constant filter function

dbaumgaertner committed:
improved treatment of time and step variables

including the correct reset of these variables in all analysis involved in the optimization

dbaumgaertner committed:
properly included damping

dbaumgaertner committed:
include possibility to specify bead direction

dbaumgaertner committed:
minor fix

dbaumgaertner committed:
added unit test

dbaumgaertner committed:
reverting performance improvements back to master state

they will be introduced in separate PRs

loumalouomega commented on applications/ShapeOptimizationApplication/python_scripts/value_logger_trust_region.py:
```diff
@@ -75,25 +75,25 @@ def _WriteCurrentValuesToFile( self ):
-            row.append(str("{:>17f}".format(self.value_history["test_norm_dX_bar"][self.current_iteration])))
-            row.append(str("{:>12f}".format(self.value_history["norm_dX"][self.current_iteration])))
-            row.append(str("{:>12f}".format(self.value_history["step_length"][self.current_iteration])))
+                row.append(" {:> .5E}".format(self.value_history[constraint_id][self.current_iteration]))
```
I hope one day we will have the table logger properly working and we could simplify this

philbucher commented on applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py:
```diff
@@ -13,6 +13,7 @@
 from __future__ import print_function, absolute_import, division
 
 # Kratos Core and Apps
+from KratosMultiphysics import *
```
I recommend to slowly shift to `import KratosMultiphysics`

philbucher commented on applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:
```diff
@@ -132,8 +146,7 @@ def __FixWholeSurface(optimization_model_part, mesh_solver_settings):
 
-        print("Add automatic process to fix the whole surface to mesh motion solver:")
-        print(auto_process_settings)
+        print("> Add automatic process to fix the whole surface to mesh motion solver:")
```
I recommend using the logger

philbucher commented on applications/ShapeOptimizationApplication/test_examples/06_bead_optimization_plate/analysis_parameters.json:
```diff
@@ -0,0 +1,108 @@
+      "processes_sub_model_part_list"      : ["corner_points","load_point"],
+      "rotation_dofs"                      : true
+  },
+  "constraints_process_list" : [{
```
the processes should be wrapped in a `processes` block
what you have now is deprecated and will eventually be removed
=> you GiD-interface seems to be quite old

dbaumgaertner replied:
@philbucher Can you post here the way it should look according to the new interface? Thanks.

philbucher replied:
where the deprecation-msg points to:https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users#analysisstage-usage

philbucher commented on applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:
```diff
@@ -595,4 +595,4 @@ inline std::ostream & operator << (std::ostream& rOStream, const BaseShellElemen
 }
 
-} // namespace Kratos.
+} // namespace Kratos.
```
please do not make such changes 
This doesn't change the code but it messes up the Git-history, which in turn makes it harder to debug

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/strain_energy_response_function_utility.h:
```diff
@@ -507,4 +507,4 @@ class StrainEnergyResponseFunctionUtility
 } // namespace Kratos.
 
-#endif // STRAIN_ENERGY_RESPONSE_FUNCTION_UTILITY_H
+#endif // STRAIN_ENERGY_RESPONSE_FUNCTION_UTILITY_H
```
idem

philbucher commented on applications/ShapeOptimizationApplication/custom_utilities/mapping/filter_function.h:
```diff
@@ -94,9 +95,13 @@ class FilterFunction
         // Throw error message in case of wrong specification
         else
-            KRATOS_ERROR << "Specified filter function type not recognized. Options are: linear , gaussian. Specified: " << std::endl;
+            KRATOS_ERROR << "Specified filter function type not recognized. Options are: constant, linear , gaussian. Specified: " << std::endl;
```
this is quite manual

alternative see [here](https://github.com/KratosMultiphysics/Kratos/blob/meshmoving/claculate-mesh-vel-utility/applications/MeshMovingApplication/custom_utilities/calculate_mesh_velocity_utility.cpp#L45-L73)

philbucher reviewed (changes requested):
please revert the changes in StrucutralMechanics, otherwise just some suggestions

dbaumgaertner committed:
included option for max total iterations

dbaumgaertner committed:
minor fix

dbaumgaertner committed:
cleanup of unused settings

dbaumgaertner committed:
Merge branch 'master' into feature-corrected-bead-optimization

dbaumgaertner committed:
minor improvement of logger

dbaumgaertner committed:
adjusting test to conform to new output of logger

dbaumgaertner committed:
removed unnecessary line

dbaumgaertner committed:
included possibility to enforce constraints without damping

constraints (fixed areas) may now be directly applied on the design variables

dbaumgaertner committed:
adjusted test to include new constraint feature

dbaumgaertner committed:
adjusted scaling

scaling is now done properly in the beginning

dbaumgaertner committed:
adjusted unit test to new scaling

dbaumgaertner committed:
slightly improved scaling again

to choose penalty factor such that it matches an estimation for the lagrange multiplier

dbaumgaertner committed:
fixed output accordingly

dbaumgaertner committed:
adjusted test to new scaling

armingeiser commented on applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h:
```diff
@@ -193,6 +205,61 @@ class GeometryUtilities
+
+                if(number_of_neighbors >= max_neighbor_nodes)
+                {
+                    std::cout << "\n> Info: For node " << node_i.Id() << " maximum number of neighbor nodes (=" << max_neighbor_nodes << " nodes) is reached! Increasing maximum number by factor 2. " << std::endl;
```
KRATOS_WARNING ?

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+            "name"                          : "bead_optimization",
+            "bead_height"                   : 1.0,
+            "bead_direction"                : [],
+            "bead_direction_mode"           : 2,
```
can we use string options that clearly name what the option means? Like e.g. "both_sides", "positive_side", "negative_side"

armingeiser commented on applications/ShapeOptimizationApplication/custom_python/shape_optimization_python_application.cpp:
```diff
@@ -85,7 +85,24 @@ PYBIND11_MODULE(KratosShapeOptimizationApplication, m)
+    KRATOS_REGISTER_IN_PYTHON_3D_VARIABLE_WITH_COMPONENTS(m, BEAD_DIRECTION);
+
+    // For auxiliary operations
+    KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, SCALAR_VARIABLE);
```
debug variables?

dbaumgaertner replied:
More like variables to be quickly used within a development process. It's always a bit annoying go through the process of adding variables just to quickly check something.

philbucher replied:
@dbaumgaertner for such cases I always use a random var from the core
There are many around :)

armingeiser replied:
:+1: for @philbucher s solution. Else please rename them to e.g. `SCALAR_DEBUG_VARIABLE` or something so it gets really clear that they are just meant for debugging.

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+        for itr in range(self.algorithm_settings["fix_boundaries"].size()):
+            sub_model_part_name = self.algorithm_settings["fix_boundaries"][itr].GetString()
+            node_set = self.optimization_model_part.GetSubModelPart(sub_model_part_name).Nodes
+            geometry_utilities.FlagNodesInRadius(node_set, IS_ALPHA_FIXED, radius)
```
after this call the ```FlagNodesInRadius``` function is not needed anymore right?

dbaumgaertner replied:
Yes.

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+        for outer_iteration in range(1,self.max_outer_iterations+1):
+            for inner_iteration in range(1,self.max_inner_iterations+1):
+
+                total_iteration = total_iteration+1
```
+=1

armingeiser reviewed (commented):
I will continue with the review tomorrow

philbucher commented on applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h:
```diff
@@ -193,6 +205,61 @@ class GeometryUtilities
+                for(unsigned int j_itr = 0 ; j_itr<number_of_neighbors ; j_itr++)
+                {
+                    ModelPart::NodeType& neighbor_node = *neighbor_nodes[j_itr];
+                    neighbor_node.SetValue(rFlagVariable,true);
```
I highly recommend using `Kratos::Flags` for this, this is much easier and intuitive to use
I can show you on Friday

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+                    for node in self.design_surface.Nodes:
+                        if not node.GetValue(IS_ALPHA_FIXED):
+                            alpha_i = node.GetSolutionStepValue(ALPHA)
+                            penalty_value = penalty_value + penalty_scaling*(alpha_i-alpha_i**2)
```
+=

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+                    for node in self.design_surface.Nodes:
+                        if not node.GetValue(IS_ALPHA_FIXED):
+                            alpha_i = node.GetSolutionStepValue(ALPHA)
+                            penalty_value = penalty_value + penalty_scaling*(-alpha_i-alpha_i**2)
```
+=

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+                    for node in self.design_surface.Nodes:
+                        if not node.GetValue(IS_ALPHA_FIXED):
+                            alpha_i = node.GetSolutionStepValue(ALPHA)
+                            penalty_value = penalty_value + penalty_scaling*(-alpha_i**2+1)
```
+=

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+                    self.mapper.InverseMap(DPDALPHA, DPDALPHA_MAPPED)
+
+                # Compute value of Lagrange function
+                L = objective_value + current_lambda*penalty_value + penalty_factor*(penalty_value)**2
```
`(penalty_value**2)` is more readable imo

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+
+                # Compute gradient of Lagrange function
+                if self.filter_penalty_term:
+                    for node in self.design_surface.Nodes:
```
this is the same loop twice, you could store `DPDALPHA` or `DPDALPHA_MAPPED` on a local variable depending on the filtering of the penalty

```python
if self.filter_penalty_term:
     _penalty_term = DF1DALPHA_MAPPED
else:
     _penalty_term = DF1DALPHA

for node in self.design_surface.Nodes:
      dLdalpha_i = node.GetSolutionStepValue(DF1DALPHA_MAPPED) + current_lambda*node.GetSolutionStepValue(_penalty_term)
      node.SetSolutionStepValue(DLDALPHA, dLdalpha_i)
```

up to you though

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+                        dLdalpha_for_normalization[node.Id] = node.GetSolutionStepValue(DLDALPHA)**2
+
+                max_value = math.sqrt(max(dLdalpha_for_normalization.values()))
+                if max_value == 0:
```
0.0, below 1.0

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+                # Map design variables
+                self.mapper.Map(ALPHA, ALPHA_MAPPED)
+                for itr in range(self.number_of_filter_iterations-1):
+                    self.mapper.Map(ALPHA_MAPPED, ALPHA_MAPPED)
```
Internally we assemble the variables into a vector, so nothing gets overwritten.
This will produce wrong results however for the matrix free mapper. (also above for the inverse mapping)

dbaumgaertner replied:
You are right. I will remove this option for multiple filtering, though. Its basically a leftover of some trials.

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py:
```diff
@@ -0,0 +1,398 @@
+                print("\n> Time needed for current optimization step = ", timer.GetLapTime(), "s")
+                print("> Time needed for total optimization so far = ", timer.GetTotalTime(), "s")
+
+            # Compute penalyt factor such that estimated Lagrange multiplier is obtained
```
```suggestion
            # Compute penalty factor such that estimated Lagrange multiplier is obtained
```

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_penalized_projection.py:
```diff
@@ -66,25 +64,36 @@ def CheckApplicability(self):
+
+        self.design_surface = self.model_part_controller.GetDesignSurface()
+
+        self.mapper = mapper_factory.CreateMapper(self.design_surface, self.design_surface, self.mapper_settings)
```
In other places in Kratos, the object (here mapper) is created in the constructor, and here only the `Initialize` function should be called. Or is there a specific reason you create the mapper in this place now?

armingeiser replied:
same for the other objects below.

dbaumgaertner replied:
The reason is, that these objects require an imported model part. And it seemed more reasonable to import the model part in InitializeOptimizationLoop, and leave the constructor with only adding the variables. That's why also these objects are created there. Actually, the objects you mention should not just expect an imported mdpa right from the beginning. But this is a bit more effort.

armingeiser replied:
I see. Importing the model part in the Initialize function is in line with other implementations in Kratos e.g. of the analysis, so :+1: . 

Please initialize the variables with `None` in constructor. Maybe we can even add a comment to remember that we should separate some code from the constructor of those classes into an Initialize function.

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/algorithm_penalized_projection.py:
```diff
@@ -109,59 +118,59 @@ def RunOptimizationLoop(self):
     def __initializeNewShape(self):
-        self.ModelPartController.UpdateMeshAccordingInputVariable(SHAPE_UPDATE)
-        self.ModelPartController.SetReferenceMeshToMesh()
+        self.model_part_controller.InitializeNewOptimizationStep(self.optimization_iteration)
```
It seems strange that in the specific function `__initializeNewShape` a more general function `InitializeNewOptimizationStep` is called. 
I would either rename the function of the model_part_controller: `InitializeNewShape` or `InitializeModelPartUpdate`
or call it in a specific function in the algorithm: `__initializeOptimizationStep`

dbaumgaertner replied:
You are right. I called it "UpdateTimeStep" now to better represent what is actually done. What do you think?

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py:
```diff
@@ -33,9 +34,19 @@ def InitializeBeforeOptimizationLoop( self ):
 
         for identifier, response in self.response_function_list.items():
 
+            # Reset step/time iterators such that they match the optimization iteration after calling CalculateValue (which internally calls CloneTimeStep)
```
:+1: if it works.

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:
```diff
@@ -54,6 +54,11 @@ def __init__(self, MeshSolverSettings, model):
         }""")
         self.MeshSolverSettings = MeshSolverSettings
         self.MeshSolverSettings.ValidateAndAssignDefaults(default_settings)
+        self.MeshSolverSettings["solver_settings"].ValidateAndAssignDefaults(default_settings["solver_settings"])
```
doesn't this line create the `mesh_motion_linear_solver_settings` if they are missing?

dbaumgaertner replied:
True. I don't know why I added this anymore.

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py:
```diff
@@ -62,10 +62,10 @@ def __init__(self, optimization_settings, model_part_controller, analyzer, commu
 
     # --------------------------------------------------------------------------
-    def __AddNodalVariablesNeededForOptimization(self):
+    def __AddGeneralVariables(self):
```
Function name is too generic, `__AddOptimizationVariables`

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py:
```diff
@@ -99,13 +98,10 @@ def Optimize(self):
         print("> ", Timer().GetTimeStamp(),": Starting optimization using the following algorithm: ", algorithm_name)
         print("> ==============================================================================================================\n")
 
-        self.model_part_controller.ImportOptimizationModelPart()
```
:+1:

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/value_logger_bead_optimization.py:
```diff
@@ -0,0 +1,77 @@
+        with open(self.complete_log_file_name, 'w') as csvfile:
+            historyWriter = csv.writer(csvfile, delimiter=',',quotechar='|',quoting=csv.QUOTE_MINIMAL)
+            row = []
+            row.append("{:>12s}".format("overall_itr"))
```
would it make sense to make all rows 13 chars long? Up to you.

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/value_logger_trust_region.py:
```diff
@@ -75,25 +75,25 @@ def _WriteCurrentValuesToFile( self ):
+            row.append("{:>12f}".format(self.value_history["bi_err"][self.current_iteration]))
+            row.append("{:>17f}".format(self.value_history["test_norm_dX_bar"][self.current_iteration]))
+            row.append("{:>12f}".format(self.value_history["norm_dX"][self.current_iteration]))
+            row.append(" {:>.5E}".format(self.value_history["step_length"][self.current_iteration]))
```
i did not check this formatting stuff, I'm fine with it if it looks good =)

armingeiser commented on applications/ShapeOptimizationApplication/shape_optimization_application.cpp:
```diff
@@ -83,6 +83,23 @@ namespace Kratos
+    KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(BEAD_DIRECTION);
+
+    // For auxiliary operations
+    KRATOS_CREATE_VARIABLE(double,SCALAR_VARIABLE);
```
unused

armingeiser commented on applications/ShapeOptimizationApplication/shape_optimization_application.cpp:
```diff
@@ -145,6 +162,23 @@ namespace Kratos
+        KRATOS_REGISTER_3D_VARIABLE_WITH_COMPONENTS(BEAD_DIRECTION);
+
+        // For auxiliary operations
+        KRATOS_REGISTER_VARIABLE(SCALAR_VARIABLE);
```
unused

armingeiser commented on applications/ShapeOptimizationApplication/shape_optimization_application.h:
```diff
@@ -85,6 +85,24 @@ namespace Kratos
+	KRATOS_DEFINE_3D_VARIABLE_WITH_COMPONENTS(BEAD_DIRECTION);
+
+    // For auxiliary operations
+    KRATOS_DEFINE_VARIABLE(double,SCALAR_VARIABLE);
```
unused

dbaumgaertner committed:
Merge branch 'master' into feature-corrected-bead-optimization

dbaumgaertner committed:
better output using kratos warning

dbaumgaertner committed:
cleaned output

dbaumgaertner committed:
algorithm improvement after review

armingeiser commented on applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h:
```diff
@@ -193,6 +205,61 @@ class GeometryUtilities
+        unsigned int bucket_size = 100;
+
+        // Create search tree
+        if(mIsSearchTreeConstructed==false)
```
I have the feeling that creating a node search as a member of the very general GeometryUtilities Class just in order to save some time when calling this function several times is a little over the top. 
The class contains many independent functionalities and the node search is only needed in this function.

Recreating the search three each time this function is called would make more sense to me.

dbaumgaertner committed:
improved naming

dbaumgaertner committed:
minor removal of unnecessary function call

dbaumgaertner committed:
adjusted tests to new algorithm design

dbaumgaertner committed:
changed to use flag instead of variable

dbaumgaertner committed:
adjusted remaining test to new output format

dbaumgaertner committed:
added control of buffer size to mdpa controller

dbaumgaertner committed:
move variable declaration to constructor

dbaumgaertner committed:
separated tree-based flagging

dbaumgaertner commented:
@armingeiser I changed the things you suggested. Can you maybe run the tests to see if everything is fine also on your computer? Thanks.

armingeiser commented on applications/ShapeOptimizationApplication/custom_utilities/tree_based_functions.h:
```diff
@@ -0,0 +1,299 @@
+
+*/
+
+class TreeBasedFunctions
```
Maybe `NodeSearchBasedFunctions` is more clear to a "user". The fact that its tree based node search is internal information.

dbaumgaertner replied:
NodeSearchBasedFunctions sounds a bit clunky. What about SearchBasedFunctions?

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:
```diff
@@ -38,10 +38,10 @@ def __init__(self, MeshSolverSettings, model):
                     "solver_type" : "AMGCL",
                     "smoother_type":"ilu0",
-                    "krylov_type": "gmres",
+                    "krylov_type": "bicgstab",
```
Is there a specific reason for this change?

dbaumgaertner replied:
No. Don't know, why this changed. I reset it.

armingeiser commented on applications/ShapeOptimizationApplication/tests/opt_process_vertex_morphing_test/run_test.py:
```diff
@@ -114,7 +114,7 @@ def __TentFunction(x):
     # Check against specifications
     TestCase().assertEqual(resulting_optimization_iterations, 124)
-    TestCase().assertAlmostEqual(resulting_improvement, -90.228425, 5)
+    TestCase().assertAlmostEqual(resulting_improvement, -90.2284, 4)
```
Did this test become more sensitive because of the current changes?

dbaumgaertner replied:
No, just the output changed.

dbaumgaertner committed:
minor renaming

dbaumgaertner committed:
forgotten change of name

armingeiser reviewed (dismissed):
My concerns have been addressed, thanks!

dbaumgaertner committed:
improved computation and output of norm

dbaumgaertner commented:
@philbucher are there still any objections from your side. Otherwise, you could approve the corrections.

@armingeiser can you still have a look at my latest commit?

philbucher commented:
just remove the changes in StructuralMechanics then I am also fine
